maitake-sync/src/wait_cell.rs

```diff
@@ -327,6 +327,159 @@ impl WaitCell {
+    /// use maitake_sync::WaitCell;
+    /// use std::sync::atomic::{AtomicU8, Ordering};
+    ///
+    /// let queue = Arc::new(WaitCell::new());
+    /// let num = Arc::new(AtomicU8::new(0));
+    ///
+    /// let waiter = task::spawn({
+    ///     // clone items to move into the spawned task
+    ///     let queue = queue.clone();
+    ///     let num = num.clone();
+    ///     async move {
+    ///         queue.wait_for(|| num.load(Ordering::Relaxed) > 5).await;
+    ///         println!("received wakeup!");
+    ///     }
+    /// });
+    ///
+    /// println!("poking task...");
+    ///
+    /// for i in 0..20 {
+    ///     num.store(i, Ordering::Relaxed);
+    ///     queue.wake();
+    /// }
+    ///
+    /// waiter.await.unwrap();
+    /// # }
+    /// # test();
+    /// ```
+    pub async fn wait_for<F: FnMut() -> bool>(&self, mut f: F) -> Result<(), Closed> {
+        loop {
+            let wait = self.subscribe().await;
+            if f() {
+                return Ok(());
+            }
+            wait.await?;
+        }
+    }
+
+    /// Asynchronously poll the given function `f` until a condition occurs,
+    /// using the [`WaitCell`] to only re-poll when notified.
+    ///
+    /// This can be used to implement a "wait loop", turning a "try" function
+    /// (e.g. "try_recv" or "try_send") into an asynchronous function (e.g.
+    /// "recv" or "send").
+    ///
+    /// In particular, this function correctly *registers* interest in the [`WaitCell`]
+    /// prior to polling the function, ensuring that there is not a chance of a race
+    /// where the condition occurs AFTER checking but BEFORE registering interest
+    /// in the [`WaitCell`], which could lead to deadlock.
+    ///
+    /// This is intended to have similar behavior to `Condvar` in the standard library,
+    /// but asynchronous, and not requiring operating system intervention (or existence).
+    ///
+    /// In particular, this can be used in cases where interrupts or events are used
+    /// to signify readiness or completion of some task, such as the completion of a
+    /// DMA transfer, or reception of an ethernet frame. In cases like this, the interrupt
+    /// can wake the cell, allowing the polling function to check status fields for
+    /// partial progress or completion, and also return the status flags at the same time.
+    ///
+    /// Consider using [`Self::wait_for()`] if your function does not return a value.
+    ///
+    /// Consider using [`WaitQueue::wait_for_value()`] if you need multiple waiters.
+    ///
+    /// * [`Ok`]`(T)` if the closure returns [`Some`]`(T)`.
+    /// * [`Err`]`(`[`Closed`]`)` if the [`WaitCell`] is closed.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # use tokio::task;
+    /// # #[tokio::main(flavor = "current_thread")]
+    /// # async fn test() {
+    /// use std::sync::Arc;
+    /// use maitake_sync::WaitCell;
+    /// use std::sync::atomic::{AtomicU8, Ordering};
```

Comment:
nit, take it or leave it: maybe the examples could get these from alloc/core just so that they "look more like" no-std code?
```suggestion
    /// use alloc::sync::Arc;
    /// use maitake_sync::WaitCell;
    /// use core::sync::atomic::{AtomicU8, Ordering};
```

Reply:
I can do this, but other tests in the same module are already using std, so I went with consistency. I can switch them all over if you'd like.

Reply:
Oh. Well, in that case, never mind!